src/azul/portal_service.py

```diff
@@ -0,0 +1,167 @@
+            if version is None:
+                log.info('Portal integration DB not found in S3; uploading hard-coded DB.')
+                self._create_db()
+                continue
+            else:
+                try:
+                    db = self._read_db(version)
+                except NoSuchObjectVersion:
+                    # Wait for latest version to appear in S3.
+                    continue
+            db = operation(db)
+            if db is None:
+                # Operation is read-only; we're done.
+                break
+            else:
+                new_version = self._write_db(db)
+                if self._update_version(version, new_version):
+                    break
+                else:
+                    # Retry with up-to-date DB
+                    continue
+
+    def _update_version(self, version: Optional[str], new_version: str) -> bool:
+        """
+        Try to update the most current version of the S3 DB and report success.
+        Update is rejected if `version` is not current.
+        :param version: the current version.
+        :param new_version: the replacement version.
+        :return: whether hte current version was updated.
+        """
+        try:
+            self.version_service.put(self._db_url(), version, new_version)
+        except VersionConflict:
+            # Operation was performed on outdated DB and now erroneously
+            # exists in S3.
+            self._delete_db(new_version)
+            return False
+        else:
+            return True
+
+    def _create_db(self) -> None:
+        """
+        Write hardcoded portal integrations DB to S3 if no other DB is present.
+        """
+        plugin = Plugin.load()
+        db = self.demultiplex(plugin.portal_integrations_db())
+        init_version = self._write_db(db)
+        self._update_version(None, init_version)
+        # Ignore failure, just means that someone else created it first
+
+    def _read_db(self, version: str) -> JSONs:
+        """
+        Retrieve specified version of portal DB from S3.
+        Raises `NoSuchKey` for missing object and `NoSuchObjectVersion` for
+        missing version.
+        """
+        try:
+            response = self.client.get_object(Bucket=config.portal_integrations_db_bucket,
+                                              Key=config.portal_integrations_db_object,
+                                              VersionId=version)
+        except self.client.exceptions.NoSuckKey:
+            # boto raises the same exception when the key is not found and when
+            # the key IS found but the version is not, so we search the object
+            # versions to disambiguate these cases.
+            all_versions = self.client.list_object_versions()['Versions']
+            if any(version == v['VersionId'] for v in all_versions):
+                raise
+            else:
+                raise NoSuchObjectVersion(version) from None
```

Comment:
It was to prevent the `NoSuchKey` exception from the surrounding `except` block from appearing at the top of the stacktrace of the `NoSuchObjectVersion` exception for clarity. No longer relevant given the previous comment.

Reply:
Generally speaking, we'd want the primary exception to be included in the trace. Suppressing information does not add clarity, except when the suppressed information would is distracting (due to volume) or misleading.